base/applications/cmdutils/taskkill/taskkill.c

```diff
@@ -237,69 +289,116 @@ static int send_close_messages(void)
             {
                 WCHAR process_name[MAX_PATH];
-
                 if (get_process_name_from_pid(pid_list[index], process_name, MAX_PATH) &&
                     !strcmpiW(process_name, task_list[i]))
                 {
-                    struct pid_close_info info = { pid_list[index] };
-
-                    found_process = TRUE;
-                    if (pid_list[index] == self_pid)
-                    {
-                        taskkill_message(STRING_SELF_TERMINATION);
-                        status_code = 1;
-                        continue;
-                    }
-
-                    EnumWindows(pid_enum_proc, (LPARAM)&info);
-                    taskkill_message_printfW(STRING_CLOSE_PROC_SRCH, process_name, pid_list[index]);
+                    pkill_list[pkill_size] = pid_list[index];
+                    pkill_size++;
                 }
             }
+        }
+
+        // Can't find any process same as name or PID
+        if (pkill_size == 0)
+        {
+            taskkill_message_printfW(STRING_SEARCH_FAILED, task_list[i]);
+            status_code = 128;
+            continue;
+        }
+
+        // Try to send close messages to process in `pkill_list`
+        for (DWORD index = 0; index < pkill_size; index++)
+        {
+            DWORD pid = pkill_list[index];
+            WCHAR process_name[MAX_PATH] = { 0 };
+            struct pid_close_info info = { pid };
 
-            if (!found_process)
+            if (pid == self_pid)
             {
-                taskkill_message_printfW(STRING_SEARCH_FAILED, task_list[i]);
-                status_code = 128;
+                taskkill_message(STRING_SELF_TERMINATION);
+                status_code = 1;
+                continue;
+            }
+
+            // Send close messages to child first
+            if (kill_child_processes)
+            {
+                send_close_messages_tree(pid);
+            }
+
+            get_process_name_from_pid(pid, process_name, MAX_PATH);
+            EnumWindows(pid_enum_proc, (LPARAM)&info);
+            if (info.found)
+            {
+                if (is_numeric)
+                {
+                    taskkill_message_printfW(STRING_CLOSE_PID_SEARCH, pid);
+                }
+                else
+                {
+                    taskkill_message_printfW(STRING_CLOSE_PROC_SRCH, process_name, pid);
+                }
             }
         }
+
+        HeapFree(GetProcessHeap(), 0, pkill_list);
     }
 
     HeapFree(GetProcessHeap(), 0, pid_list);
     return status_code;
 }
 
+static void terminate_process_tree(DWORD ppid)
+{
+    DWORD child_size = 0, index = 0;
+    DWORD* child_pid_list = find_child_pid(ppid, &child_size);
+
+    // Use recursion to browse all child processes
+    while (index < child_size)
+    {
+        terminate_process_tree(child_pid_list[index]);
+        HANDLE process = OpenProcess(PROCESS_TERMINATE, FALSE, child_pid_list[index]);
```

Comment:
Another variable declaration that will need to move.

Reply:
Thanks, I fixed.